File: src/silicadbd.c
#include <errno.h>
#include <fcntl.h>
#include <poll.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/un.h>
#include <unistd.h>

#include "proto.h"
#include "store.h"
#include "wire.h"

#define MAXC 64

typedef struct {
    int fd;
    buf_t in, out;
} conn_t;

static volatile sig_atomic_t g_stop;
static void on_sig(int s) { (void)s; g_stop = 1; }

static store_t *g_st;

static void respond(conn_t *c, uint8_t op, uint64_t rid, uint16_t status, const buf_t *pl) {
    uint8_t h[SLDB_HDR_SIZE];
    uint32_t n = pl ? (uint32_t)pl->len : 0;
    hdr_write(h, n, op, F_RESP, status, rid);
    buf_put(&c->out, h, sizeof h);
    if (n) buf_put(&c->out, pl->p, n);
}

static void list_cb(const char *key, uint8_t kind, uint64_t ts, void *ud) {
    buf_t *b = ud;
    buf_tlv_str(b, T_KEY, key);
    buf_tlv_u8(b, T_KIND, kind);
    buf_tlv_u64(b, T_TS, ts);
}

static void links_cb(const char *s, const char *p, const char *o, uint64_t ts, void *ud) {
    buf_t *b = ud;
    buf_tlv_str(b, T_SUBJ, s);
    buf_tlv_str(b, T_PRED, p);
    buf_tlv_str(b, T_OBJ, o);
    buf_tlv_u64(b, T_TS, ts);
}

static void dispatch(conn_t *c, uint8_t op, uint64_t rid, const uint8_t *pl, uint32_t n) {
    buf_t r;
    buf_init(&r);
    uint16_t st = ST_OK;
    char key[SLDB_KEY_MAX + 1];

    switch (op) {
    case OP_HELLO: {
        const uint8_t *v;
        uint32_t vl, ver = 0;
        if (tlv_find(pl, n, T_VERSION, &v, &vl) == 1 && vl == 4) ver = le32r(v);
        if (ver != SLDB_VERSION) st = ST_VERSION;
        buf_tlv_u32(&r, T_VERSION, SLDB_VERSION);
        break;
    }
    case OP_PING:
        break;
    case OP_PUT: {
        uint64_t ts;
        int rc;
        if (tlv_find_u64(pl, n, T_TS, &ts) == 1) {
            rc = store_put(g_st, pl, n);
        } else {
            buf_t p2;
            buf_init(&p2);
            if (buf_put(&p2, pl, n) || buf_tlv_u64(&p2, T_TS, now_ns()))
                rc = -1;
            else
                rc = store_put(g_st, p2.p, (uint32_t)p2.len);
            buf_free(&p2);
        }
        st = rc == 0 ? ST_OK : rc == -2 ? ST_BADREQ : ST_IO;
        break;
    }
    case OP_GET: {
        if (tlv_find_str(pl, n, T_KEY, key, sizeof key) != 1 || !key[0]) {
            st = ST_BADREQ;
            break;
        }
        uint8_t *v;
        uint32_t vl;
        int rc = store_get(g_st, key, &v, &vl);
        if (rc == 1) {
            buf_put(&r, v, vl);
            free(v);
        } else {
            st = rc == 0 ? ST_NOTFOUND : ST_IO;
        }
        break;
    }
    case OP_DEL: {
        if (tlv_find_str(pl, n, T_KEY, key, sizeof key) != 1 || !key[0]) {
            st = ST_BADREQ;
            break;
        }
        int rc = store_del(g_st, key);
        st = rc == 1 ? ST_OK : rc == 0 ? ST_NOTFOUND : ST_IO;
        break;
    }
    case OP_LIST: {
        char pfx[SLDB_KEY_MAX + 1] = "";
        if (tlv_find_str(pl, n, T_PREFIX, pfx, sizeof pfx) < 0) {
            st = ST_BADREQ;
            break;
        }
        store_iter_keys(g_st, pfx[0] ? pfx : NULL, list_cb, &r);
        break;
    }
    case OP_LINK: {
        char s[SLDB_KEY_MAX + 1], p[SLDB_KEY_MAX + 1], o[SLDB_KEY_MAX + 1];
        if (tlv_find_str(pl, n, T_SUBJ, s, sizeof s) != 1 ||
            tlv_find_str(pl, n, T_PRED, p, sizeof p) != 1 ||
            tlv_find_str(pl, n, T_OBJ, o, sizeof o) != 1 ||
            !s[0] || !p[0] || !o[0]) {
            st = ST_BADREQ;
            break;
        }
        if (store_link(g_st, s, p, o, now_ns())) st = ST_IO;
        break;
    }
    case OP_LINKS: {
        key[0] = 0;
        if (tlv_find_str(pl, n, T_KEY, key, sizeof key) < 0) {
            st = ST_BADREQ;
            break;
        }
        store_iter_links(g_st, key[0] ? key : NULL, links_cb, &r);
        break;
    }
    case OP_STATS:
        buf_tlv_u64(&r, T_NKEYS, store_nkeys(g_st));
        buf_tlv_u64(&r, T_NLINKS, store_nlinks(g_st));
        buf_tlv_u64(&r, T_BYTES, store_bytes(g_st));
        break;
    default:
        st = ST_BADREQ;
        buf_tlv_str(&r, T_MSG, "unknown op");
        break;
    }

    respond(c, op, rid, st, &r);
    buf_free(&r);
}

/* 0 = ok, -1 = close connection */
static int conn_process(conn_t *c) {
    for (;;) {
        if (c->in.len < SLDB_HDR_SIZE) return 0;
        uint32_t len;
        uint8_t op, fl;
        uint16_t st;
        uint64_t rid;
        hdr_read(c->in.p, &len, &op, &fl, &st, &rid);
        if (len > SLDB_MAX_PAYLOAD) return -1;
        if (c->in.len < SLDB_HDR_SIZE + (size_t)len) return 0;
        dispatch(c, op, rid, c->in.p + SLDB_HDR_SIZE, len);
        buf_consume(&c->in, SLDB_HDR_SIZE + len);
    }
}

static int conn_read(conn_t *c) {
    uint8_t tmp[65536];
    int eof = 0;
    for (;;) {
        ssize_t r = read(c->fd, tmp, sizeof tmp);
        if (r > 0) {
            if (buf_put(&c->in, tmp, (size_t)r)) return -1;
            if (r < (ssize_t)sizeof tmp) break;
            continue;
        }
        if (r == 0) {
            eof = 1;
            break;
        }
        if (errno == EAGAIN || errno == EWOULDBLOCK) break;
        if (errno == EINTR) continue;
        return -1;
    }
    if (conn_process(c)) return -1;
    return eof ? -1 : 0;
}

static int conn_flush(conn_t *c) {
    while (c->out.len) {
        ssize_t w = write(c->fd, c->out.p, c->out.len);
        if (w > 0) {
            buf_consume(&c->out, (size_t)w);
            continue;
        }
        if (errno == EAGAIN || errno == EWOULDBLOCK) return 0;
        if (errno == EINTR) continue;
        return -1;
    }
    return 0;
}

static void conn_close(conn_t *c) {
    conn_flush(c); /* best effort */
    close(c->fd);
    c->fd = -1;
    buf_free(&c->in);
    buf_free(&c->out);
}

int main(int argc, char **argv) {
    const char *dir = NULL;
    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "-d") && i + 1 < argc) {
            dir = argv[++i];
        } else {
            fprintf(stderr, "usage: silicadbd [-d dir]\n");
            return 1;
        }
    }

    char home[512];
    if (dir) {
        snprintf(home, sizeof home, "%s", dir);
    } else {
        const char *env = getenv("SILICADB_HOME");
        if (env && *env) {
            snprintf(home, sizeof home, "%s", env);
        } else {
            const char *h = getenv("HOME");
            if (!h) {
                fprintf(stderr, "silicadbd: HOME unset\n");
                return 1;
            }
            snprintf(home, sizeof home, "%s/.silicadb", h);
        }
    }
    mkdir(home, 0700);

    char logp[600], sockp[600];
    snprintf(logp, sizeof logp, "%s/memory.log", home);
    snprintf(sockp, sizeof sockp, "%s/silicadb.sock", home);

    g_st = store_open(logp);
    if (!g_st) {
        fprintf(stderr, "silicadbd: cannot open %s: %s\n", logp, strerror(errno));
        return 1;
    }

    struct sockaddr_un sa;
    if (strlen(sockp) >= sizeof sa.sun_path) {
        fprintf(stderr, "silicadbd: socket path too long: %s\n", sockp);
        return 1;
    }
    int lfd = socket(AF_UNIX, SOCK_STREAM, 0);
    if (lfd < 0) {
        perror("silicadbd: socket");
        return 1;
    }
    memset(&sa, 0, sizeof sa);
    sa.sun_family = AF_UNIX;
    strcpy(sa.sun_path, sockp);
    unlink(sockp);
    if (bind(lfd, (struct sockaddr *)&sa, sizeof sa) || listen(lfd, 64)) {
        perror("silicadbd: bind/listen");
        return 1;
    }
    fcntl(lfd, F_SETFL, O_NONBLOCK);

    signal(SIGPIPE, SIG_IGN);
    struct sigaction sact;
    memset(&sact, 0, sizeof sact);
    sact.sa_handler = on_sig;
    sigaction(SIGINT, &sact, NULL);
    sigaction(SIGTERM, &sact, NULL);

    fprintf(stderr, "silicadbd: %llu keys, %llu links; listening on %s\n",
            (unsigned long long)store_nkeys(g_st),
            (unsigned long long)store_nlinks(g_st), sockp);

    conn_t conns[MAXC];
    for (int i = 0; i < MAXC; i++) conns[i].fd = -1;

    while (!g_stop) {
        struct pollfd pf[MAXC + 1];
        int map[MAXC + 1], np = 0;
        pf[np].fd = lfd;
        pf[np].events = POLLIN;
        pf[np].revents = 0;
        np++;
        for (int i = 0; i < MAXC; i++) {
            if (conns[i].fd < 0) continue;
            pf[np].fd = conns[i].fd;
            pf[np].events = (short)(POLLIN | (conns[i].out.len ? POLLOUT : 0));
            pf[np].revents = 0;
            map[np] = i;
            np++;
        }
        if (poll(pf, (nfds_t)np, -1) < 0) {
            if (errno == EINTR) continue;
            perror("silicadbd: poll");
            break;
        }
        if (pf[0].revents & POLLIN) {
            for (;;) {
                int fd = accept(lfd, NULL, NULL);
                if (fd < 0) break;
                fcntl(fd, F_SETFL, O_NONBLOCK);
                int slot = -1;
                for (int i = 0; i < MAXC; i++)
                    if (conns[i].fd < 0) { slot = i; break; }
                if (slot < 0) {
                    close(fd);
                    continue;
                }
                conns[slot].fd = fd;
                buf_init(&conns[slot].in);
                buf_init(&conns[slot].out);
            }
        }
        for (int j = 1; j < np; j++) {
            conn_t *c = &conns[map[j]];
            if (c->fd < 0) continue;
            int dead = 0;
            if (pf[j].revents & (POLLERR | POLLNVAL)) dead = 1;
            if (!dead && (pf[j].revents & (POLLIN | POLLHUP))) dead = conn_read(c) < 0;
            if (!dead && c->out.len) dead = conn_flush(c) < 0;
            if (dead) conn_close(c);
        }
    }

    for (int i = 0; i < MAXC; i++)
        if (conns[i].fd >= 0) conn_close(&conns[i]);
    close(lfd);
    unlink(sockp);
    store_close(g_st);
    fprintf(stderr, "silicadbd: bye\n");
    return 0;
}

File: src/silicadbd.zig
//! silicadbd — memory daemon. poll(2) loop over a unix stream socket.

const std = @import("std");
const c = std.c;
const proto = @import("proto.zig");
const wire = @import("wire.zig");
const store_mod = @import("store.zig");
const Store = store_mod.Store;

const gpa = std.heap.c_allocator;
const MAXC = 64;

extern "c" fn signal(sig: c_int, handler: ?*const fn (c_int) callconv(.c) void) ?*const fn (c_int) callconv(.c) void;
fn sigNop(_: c_int) callconv(.c) void {}
const SIGINT: c_int = 2;
const SIGPIPE: c_int = 13;
const SIGTERM: c_int = 15;

var g_stop: bool = false;
fn onSig(_: c_int) callconv(.c) void {
    g_stop = true;
}

var g_st: Store = undefined;

const Conn = struct {
    fd: c.fd_t = -1,
    in: std.ArrayList(u8) = .empty,
    out: std.ArrayList(u8) = .empty,
};

fn errPrint(comptime fmt: []const u8, args: anytype) void {
    var b: [1024]u8 = undefined;
    const s = std.fmt.bufPrint(&b, fmt, args) catch return;
    wire.writeFull(2, s) catch {};
}

fn respond(conn: *Conn, op: u8, rid: u64, status: u16, pl: []const u8) void {
    var h: [proto.HDR_SIZE]u8 = undefined;
    wire.hdrWrite(&h, .{ .len = @intCast(pl.len), .op = op, .flags = proto.F_RESP, .status = status, .rid = rid });
    conn.out.appendSlice(gpa, &h) catch {};
    conn.out.appendSlice(gpa, pl) catch {};
}

fn stFromErr(e: Store.Error) u16 {
    return switch (e) {
        error.BadPayload => proto.ST_BADREQ,
        error.Io, error.OutOfMemory => proto.ST_IO,
    };
}

fn dispatch(conn: *Conn, op: u8, rid: u64, pl: []const u8) void {
    var r: std.ArrayList(u8) = .empty;
    defer r.deinit(gpa);
    var status: u16 = proto.ST_OK;

    blk: {
        switch (op) {
            proto.OP_HELLO => {
                var ver: u32 = 0;
                if (wire.find(pl, proto.T_VERSION) catch null) |v| {
                    if (v.len == 4) ver = std.mem.readInt(u32, v[0..4], .little);
                }
                if (ver != proto.VERSION) status = proto.ST_VERSION;
                wire.tlvU32(&r, gpa, proto.T_VERSION, proto.VERSION) catch {};
            },
            proto.OP_PING => {},
            proto.OP_PUT => {
                const has_ts = wire.findU64(pl, proto.T_TS) catch {
                    status = proto.ST_BADREQ;
                    break :blk;
                };
                if (has_ts != null) {
                    g_st.put(pl) catch |e| {
                        status = stFromErr(e);
                    };
                } else {
                    var p2: std.ArrayList(u8) = .empty;
                    defer p2.deinit(gpa);
                    p2.appendSlice(gpa, pl) catch {
                        status = proto.ST_IO;
                        break :blk;
                    };
                    wire.tlvU64(&p2, gpa, proto.T_TS, wire.nowNs()) catch {
                        status = proto.ST_IO;
                        break :blk;
                    };
                    g_st.put(p2.items) catch |e| {
                        status = stFromErr(e);
                    };
                }
            },
            proto.OP_GET => {
                const key = (wire.findStr(pl, proto.T_KEY, proto.KEY_MAX) catch null) orelse {
                    status = proto.ST_BADREQ;
                    break :blk;
                };
                if (key.len == 0) {
                    status = proto.ST_BADREQ;
                    break :blk;
                }
                const stored = g_st.get(gpa, key) catch |e| {
                    status = stFromErr(e);
                    break :blk;
                };
                if (stored) |s| {
                    defer gpa.free(s);
                    r.appendSlice(gpa, s) catch {
                        status = proto.ST_IO;
                    };
                } else {
                    status = proto.ST_NOTFOUND;
                }
            },
            proto.OP_DEL => {
                const key = (wire.findStr(pl, proto.T_KEY, proto.KEY_MAX) catch null) orelse {
                    status = proto.ST_BADREQ;
                    break :blk;
                };
                if (key.len == 0) {
                    status = proto.ST_BADREQ;
                    break :blk;
                }
                const deleted = g_st.del(key) catch |e| {
                    status = stFromErr(e);
                    break :blk;
                };
                if (!deleted) status = proto.ST_NOTFOUND;
            },
            proto.OP_LIST => {
                const pfx = wire.findStr(pl, proto.T_PREFIX, proto.KEY_MAX) catch {
                    status = proto.ST_BADREQ;
                    break :blk;
                };
                var it = g_st.idx.iterator();
                while (it.next()) |e| {
                    const key = e.key_ptr.*;
                    if (pfx) |p| {
                        if (!std.mem.startsWith(u8, key, p)) continue;
                    }
                    wire.tlv(&r, gpa, proto.T_KEY, key) catch break;
                    wire.tlvU8(&r, gpa, proto.T_KIND, e.value_ptr.kind) catch break;
                    wire.tlvU64(&r, gpa, proto.T_TS, e.value_ptr.ts) catch break;
                }
            },
            proto.OP_LINK => {
                const s = (wire.findStr(pl, proto.T_SUBJ, proto.KEY_MAX) catch null) orelse "";
                const p = (wire.findStr(pl, proto.T_PRED, proto.KEY_MAX) catch null) orelse "";
                const o = (wire.findStr(pl, proto.T_OBJ, proto.KEY_MAX) catch null) orelse "";
                if (s.len == 0 or p.len == 0 or o.len == 0) {
                    status = proto.ST_BADREQ;
                    break :blk;
                }
                const w = (wire.findF32(pl, proto.T_WEIGHT) catch {
                    status = proto.ST_BADREQ;
                    break :blk;
                }) orelse 1.0;
                const src = (wire.findStr(pl, proto.T_SRC, proto.SRC_MAX) catch {
                    status = proto.ST_BADREQ;
                    break :blk;
                }) orelse "";
                g_st.link(s, p, o, w, src, wire.nowNs()) catch |e| {
                    status = stFromErr(e);
                };
            },
            proto.OP_LINKS => {
                const key = wire.findStr(pl, proto.T_KEY, proto.KEY_MAX) catch {
                    status = proto.ST_BADREQ;
                    break :blk;
                };
                for (g_st.links.items) |l| {
                    if (key) |k| {
                        if (!std.mem.eql(u8, l.s, k) and !std.mem.eql(u8, l.o, k)) continue;
                    }
                    wire.tlv(&r, gpa, proto.T_SUBJ, l.s) catch break;
                    wire.tlv(&r, gpa, proto.T_PRED, g_st.predName(l.pid)) catch break;
                    wire.tlv(&r, gpa, proto.T_OBJ, l.o) catch break;
                    wire.tlvF32(&r, gpa, proto.T_WEIGHT, l.w) catch break;
                    if (l.src.len > 0) wire.tlv(&r, gpa, proto.T_SRC, l.src) catch break;
                    wire.tlvU64(&r, gpa, proto.T_TS, l.ts) catch break;
                }
            },
            proto.OP_STATS => {
                wire.tlvU64(&r, gpa, proto.T_NKEYS, g_st.nkeys()) catch {};
                wire.tlvU64(&r, gpa, proto.T_NLINKS, g_st.nlinks()) catch {};
                wire.tlvU64(&r, gpa, proto.T_BYTES, g_st.bytes()) catch {};
            },
            else => {
                status = proto.ST_BADREQ;
                wire.tlv(&r, gpa, proto.T_MSG, "unknown op") catch {};
            },
        }
    }

    respond(conn, op, rid, status, r.items);
}

/// false = close connection
fn connProcess(conn: *Conn) bool {
    while (true) {
        if (conn.in.items.len < proto.HDR_SIZE) return true;
        const hdr = wire.hdrRead(conn.in.items[0..proto.HDR_SIZE]);
        if (hdr.len > proto.MAX_PAYLOAD) return false;
        const total = proto.HDR_SIZE + @as(usize, hdr.len);
        if (conn.in.items.len < total) return true;
        dispatch(conn, hdr.op, hdr.rid, conn.in.items[proto.HDR_SIZE..total]);
        std.mem.copyForwards(u8, conn.in.items[0 .. conn.in.items.len - total], conn.in.items[total..]);
        conn.in.items.len -= total;
    }
}

fn connRead(conn: *Conn) bool {
    var tmp: [65536]u8 = undefined;
    var eof = false;
    while (true) {
        const r = c.read(conn.fd, &tmp, tmp.len);
        if (r > 0) {
            conn.in.appendSlice(gpa, tmp[0..@intCast(r)]) catch return false;
            if (r < tmp.len) break;
            continue;
        }
        if (r == 0) {
            eof = true;
            break;
        }
        const e = c.errno(r);
        if (e == .AGAIN) break;
        if (e == .INTR) continue;
        return false;
    }
    if (!connProcess(conn)) return false;
    return !eof;
}

fn connFlush(conn: *Conn) bool {
    while (conn.out.items.len > 0) {
        const w = c.write(conn.fd, conn.out.items.ptr, conn.out.items.len);
        if (w > 0) {
            const n: usize = @intCast(w);
            std.mem.copyForwards(u8, conn.out.items[0 .. conn.out.items.len - n], conn.out.items[n..]);
            conn.out.items.len -= n;
            continue;
        }
        const e = c.errno(w);
        if (e == .AGAIN) return true;
        if (e == .INTR) continue;
        return false;
    }
    return true;
}

fn connClose(conn: *Conn) void {
    _ = connFlush(conn); // best effort
    _ = c.close(conn.fd);
    conn.fd = -1;
    conn.in.deinit(gpa);
    conn.out.deinit(gpa);
    conn.in = .empty;
    conn.out = .empty;
}

fn setNonblock(fd: c.fd_t) void {
    const nb: c_int = @intCast(@as(u32, @bitCast(c.O{ .NONBLOCK = true })));
    _ = c.fcntl(fd, c.F.SETFL, nb);
}

fn getenvSlice(name: [*:0]const u8) ?[]const u8 {
    const v = c.getenv(name) orelse return null;
    return std.mem.span(v);
}

pub export fn main(argc: c_int, argv: [*][*:0]u8) c_int {
    var dir: ?[]const u8 = null;
    var i: usize = 1;
    const n: usize = @intCast(argc);
    while (i < n) : (i += 1) {
        const a = std.mem.span(argv[i]);
        if (std.mem.eql(u8, a, "-d") and i + 1 < n) {
            i += 1;
            dir = std.mem.span(argv[i]);
        } else {
            errPrint("usage: silicadbd [-d dir]\n", .{});
            return 1;
        }
    }

    var homebuf: [512]u8 = undefined;
    const home: []const u8 = dir orelse getenvSlice("SILICADB_HOME") orelse blk: {
        const h = getenvSlice("HOME") orelse {
            errPrint("silicadbd: HOME unset\n", .{});
            return 1;
        };
        break :blk std.fmt.bufPrint(&homebuf, "{s}/.silicadb", .{h}) catch return 1;
    };

    var logbuf: [600]u8 = undefined;
    var sockbuf: [600]u8 = undefined;
    const logp = std.fmt.bufPrintZ(&logbuf, "{s}/memory.log", .{home}) catch return 1;
    const sockp = std.fmt.bufPrintZ(&sockbuf, "{s}/silicadb.sock", .{home}) catch return 1;

    var homez: [512]u8 = undefined;
    const homeZ = std.fmt.bufPrintZ(&homez, "{s}", .{home}) catch return 1;
    _ = c.mkdir(homeZ.ptr, 0o700);

    g_st = Store.open(gpa, logp) catch {
        errPrint("silicadbd: cannot open {s}\n", .{logp});
        return 1;
    };
    defer g_st.close();

    var sa: c.sockaddr.un = .{ .path = @splat(0) };
    if (sockp.len >= sa.path.len) {
        errPrint("silicadbd: socket path too long: {s}\n", .{sockp});
        return 1;
    }
    const lfd = c.socket(c.AF.UNIX, c.SOCK.STREAM, 0);
    if (lfd < 0) {
        errPrint("silicadbd: socket failed\n", .{});
        return 1;
    }
    @memcpy(sa.path[0..sockp.len], sockp);
    _ = c.unlink(sockp.ptr);
    if (c.bind(lfd, @ptrCast(&sa), @sizeOf(c.sockaddr.un)) != 0 or c.listen(lfd, 64) != 0) {
        errPrint("silicadbd: bind/listen failed on {s}\n", .{sockp});
        return 1;
    }
    setNonblock(lfd);

    _ = signal(SIGPIPE, sigNop);
    _ = signal(SIGINT, onSig);
    _ = signal(SIGTERM, onSig);

    errPrint("silicadbd: {d} keys, {d} links; listening on {s}\n", .{ g_st.nkeys(), g_st.nlinks(), sockp });

    var conns: [MAXC]Conn = @splat(.{});

    while (!g_stop) {
        var pf: [MAXC + 1]c.pollfd = undefined;
        var map: [MAXC + 1]usize = undefined;
        var np: usize = 1;
        pf[0] = .{ .fd = lfd, .events = c.POLL.IN, .revents = 0 };
        for (&conns, 0..) |*conn, ci| {
            if (conn.fd < 0) continue;
            var ev: i16 = c.POLL.IN;
            if (conn.out.items.len > 0) ev |= c.POLL.OUT;
            pf[np] = .{ .fd = conn.fd, .events = ev, .revents = 0 };
            map[np] = ci;
            np += 1;
        }
        const pr = c.poll(&pf, @intCast(np), -1);
        if (pr < 0) {
            if (c.errno(pr) == .INTR) continue;
            errPrint("silicadbd: poll failed\n", .{});
            break;
        }
        if (pf[0].revents & c.POLL.IN != 0) {
            while (true) {
                const fd = c.accept(lfd, null, null);
                if (fd < 0) break;
                setNonblock(fd);
                var slot: ?usize = null;
                for (&conns, 0..) |*conn, ci| {
                    if (conn.fd < 0) {
                        slot = ci;
                        break;
                    }
                }
                if (slot) |sl| {
                    conns[sl] = .{ .fd = fd };
                } else {
                    _ = c.close(fd);
                }
            }
        }
        for (pf[1..np], map[1..np]) |*p, ci| {
            const conn = &conns[ci];
            if (conn.fd < 0) continue;
            var alive = true;
            if (p.revents & (c.POLL.ERR | c.POLL.NVAL) != 0) alive = false;
            if (alive and p.revents & (c.POLL.IN | c.POLL.HUP) != 0) alive = connRead(conn);
            if (alive and conn.out.items.len > 0) alive = connFlush(conn);
            if (!alive) connClose(conn);
        }
    }

    for (&conns) |*conn| {
        if (conn.fd >= 0) connClose(conn);
    }
    _ = c.close(lfd);
    _ = c.unlink(sockp.ptr);
    errPrint("silicadbd: bye\n", .{});
    return 0;
}
